packages/inquirer/src/types.mts

```diff
@@ -1,88 +1,93 @@
+ */
 export interface QuestionMap {
-  input: Parameters<typeof input>[0];
-  select: Parameters<typeof select>[0];
-  /** @deprecated `list` is now named `select` */
-  list: Parameters<typeof select>[0];
-  number: Parameters<typeof number>[0];
-  confirm: Parameters<typeof confirm>[0];
-  rawlist: Parameters<typeof rawlist>[0];
-  expand: Parameters<typeof expand>[0];
-  checkbox: Parameters<typeof checkbox>[0];
-  password: Parameters<typeof password>[0];
-  editor: Parameters<typeof editor>[0];
+  // Dummy key to avoid empty object type
+  __dummy: { message: string };
 }
 
-type PromptConfigMap<A extends Answers> = {
-  [key in keyof QuestionMap]: Readonly<
-    DistributiveMerge<
-      QuestionMap[keyof QuestionMap],
-      {
-        type: keyof QuestionMap;
-        name: KeyUnion<A>;
-        when?: AsyncGetterFunction<boolean, Prettify<A>> | boolean;
-        askAnswered?: boolean;
-        message:
-          | Pick<QuestionMap[keyof QuestionMap], 'message'>
-          | AsyncGetterFunction<
-              Pick<QuestionMap[keyof QuestionMap], 'message'>,
-              Prettify<A>
-            >;
-        choices?:
-          | Pick<QuestionMap[keyof QuestionMap], 'choices'>
-          | string[]
-          | AsyncGetterFunction<
-              Pick<QuestionMap[keyof QuestionMap], 'choices'> | string[],
-              Prettify<A>
-            >;
-        default?:
-          | Pick<QuestionMap[keyof QuestionMap], 'default'>
-          | AsyncGetterFunction<
-              Pick<QuestionMap[keyof QuestionMap], 'default'> | string[],
-              Prettify<A>
-            >;
-      }
-    >
-  >;
+type KeyValueOrAsyncGetterFunction<T, k extends string, A extends Answers> =
+  T extends Record<string, any> ? T[k] | AsyncGetterFunction<T[k], A> : never;
+
+export type Named<T, N = string> = T & { name: N };
+
+export type LegacyQuestion<A extends Answers, Type extends string = string> = {
+  type: Type;
+  askAnswered?: boolean;
+  when?: boolean | AsyncGetterFunction<boolean, A>;
 };
 
-export type Question<A extends Answers> = PromptConfigMap<A>[keyof PromptConfigMap<A>];
+export type NamedLegacyQuestion<A extends Answers, Type extends string = string> = Named<
+  LegacyQuestion<A, Type>
+>;
 
-export type QuestionAnswerMap<A extends Answers> = Readonly<{
-  [name in KeyUnion<A>]: Omit<Question<A>, 'name'>;
-}>;
+export type LegacyAsyncQuestion<
+  Type extends string,
+  Q extends Record<string, any>,
+  A extends Answers,
+> = DistributiveMerge<
+  Q,
+  LegacyQuestion<A, Type> & {
+    filter?(input: any, answers: A): any;
+    message: KeyValueOrAsyncGetterFunction<Q, 'message', A>;
+    default?: KeyValueOrAsyncGetterFunction<Q, 'default', A>;
+    choices?: KeyValueOrAsyncGetterFunction<Q, 'choices', A>;
+  }
+>;
 
-export type QuestionArray<A extends Answers> = readonly Question<A>[];
+export type Question<A extends Answers = object> =
+  | LegacyAsyncQuestion<'confirm', ConfirmConfig, A>
```

Comment:
Why export/import the types instead of inferring them?

e.g. more similar to how it was before

```diff
- LegacyAsyncQuestion<'confirm', ConfirmConfig, A>
- LegacyAsyncQuestion<'confirm', Parameters<typeof confirm>[0], A>
```

Reply:
I gave it a quick try and it seemed to work just fine.

I strongly prefer the second format as I don't want to tie the library to the internal type implementation. This grows the potential breaking change surface, and makes our public API larger.